### DISCUSSION TITLE
Custom modify previewer title on title overflow

### DISCUSSION THREAD

**Discussion by BSteffaniak** [ibhagwan/fzf-lua#1132] [Ideas]
Hi. This might already be able to be achieved somehow, but I wasn't able to achieve it without modifying the source code myself.

I want to handle shortening an overflowed preview title by prepending a `...` to the front of the title and chopping off the left part of the title.

e.g. if the title `this is a test title` only accepts 15 characters, it would shorten to `...a test title`

The way I handled this was to add an optional title_fnameoverflowmodify transformer to winopts.preview. I initially tried to add that to previewers.builtin, but I don't know how to access that from the win.lua#FzfWin:update_title function. If I could've found a way to access that in that function, I probably would have added it there. However, I'm not very familiar with lua, so I just added it to winopts.preview since that function was already using stuff from that.

I then used that title_fnameoverflowmodify in the update_title function like this:
```lua
if #title > width - right_pad then
  if self.winopts.preview.title_fnameoverflowmodify then
    title = self.winopts.preview.title_fnameoverflowmodify(title, width - right_pad)
  end
  title = title:sub(1, width - right_pad) .. " " -- still chop off any overflow, if title_fnameoverflowmodify didn't properly truncate it
end
```

With this, I then defined winopts.preview.title_fnameoverflowmodify in my setup config like this:

```lua
winopts = {
  preview = {
    title_fnameoverflowmodify = function(t, len)
      return " ..." .. t:sub(#t - len + 3 + 1, #t)
    end,
  },
},
```

The result looks like this:

![image](https://github.com/ibhagwan/fzf-lua/assets/1524311/9f82efa8-2ea2-4530-a7f8-e19e9608b894)

Is this achievable in a different -- less complex -- way? If not, does this seem like a good solution?

Thanks!

**Comment by ibhagwan** [comment]
That’s a good solution @BSteffaniak, the only exception is that I would prefer to avoid yet another obscure function, instead we can just add width as the second parameter to the already existing `title_fnamemodify`?

**Comment by ibhagwan** [comment]
https://github.com/ibhagwan/fzf-lua/commit/344b309421e5222a6199e4b46d01041089b6a2ae - this adds `width` as the 2nd parameter to the already existing `title_fnamemodify`, since `right_pad` is constant you can still achieve the same result.

**Reply by BSteffaniak** [reply]
Awesome! thanks!

**Reply by BSteffaniak** [reply]
Is there any way to get the left/right padding values? The width being passed seems to be the whole top border width.

**Reply by ibhagwan** [reply]
These are hardcoded (number of border corner chars), you can copy the value from the code (7 if I recall correctly).

**Reply by BSteffaniak** [reply]
This is what I ended up with that works for me:

```lua
previewers = {
  builtin = {
    title_fnamemodify = function(t, width)
      local min_left_padding = 4
      local min_right_padding = 4
      local max_text_width = width - min_left_padding - min_right_padding
      if #t > max_text_width then
        return "..." .. t:sub(#t - max_text_width + 3 + 1, #t)
      end
      return t
    end,
  },
},
```

**Reply by ibhagwan** [reply]
Looks great!
